### PR TITLE
Codechange: let FormatString use StringBuilder

### PR DESCRIPTION
**Pull request by rubidium42**
## Motivation / Problem

The strings system uses a (limited) char buffer to write the data strings to. This has to be changed to std::string.

## Description

Use the `StringBuilder` in `FormatString`, add some migration helpers so `GetStringWithArgs` can still remain as it is.

## Limitations

Depends on #10884 to get merged.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/strings.cpp [hunk 1]
```diff
@@ -281,6 +289,20 @@ char *GetString(char *buffr, StringID string, const char *last)
 	return GetStringWithArgs(buffr, string, &_global_string_params, last);
 }
 
+/**
+ * Migration helper to get a parsed string with most special stringcodes replaced by the string parameters.
```
Comment (TrueBrain) [inline src/strings.cpp]: I would leave out "Migration helper" .. who knows, maybe you get bored tomorrow and never continue. Which is fine. But then this comment makes little sense :D

File: src/strings.cpp [hunk 2]
```diff
@@ -806,18 +828,18 @@ static std::vector<const char *> _game_script_raw_strings;
...
  * @param dry_run True when the argt array is not yet initialized.
  */
-static char *FormatString(char *buff, const char *str_arg, StringParameters *args, const char *last, uint case_index, bool game_script, bool dry_run)
+static void FormatString(StringBuilder &builder, const char *str_arg, StringParameters *args, uint case_index, bool game_script, bool dry_run)
 {
 	uint orig_offset = args->offset;
 
 	/* When there is no array with types there is no need to do a dry run. */
 	if (args->HasTypeInformation() && !dry_run) {
+		StringBuilder dry_run_builder = builder;
```
Comment (TrueBrain) [inline src/strings.cpp:850]: Codewise this feels off. The line above says `if not dry-run`, after which you make a `dry-run` variable. But also: why? The old code didn't do it?

As you can see, I don't actually understand the code, so it might be easily explained :)

File: src/strings.cpp [hunk 3]
```diff
@@ -981,31 +1003,31 @@ static char *FormatString(char *buff, const char *str_arg, StringParameters *arg
 					char *p = input + Utf8Encode(input, args->GetTypeAtOffset(offset));
 					*p = '\0';
 
-					/* Now do the string formatting. */
-					char buf[256];
+					/* Now do the string formatting; we need the first character, so that plus '\0' is enough. */
```
Comment (TrueBrain) [inline src/strings.cpp]: "we need the first character"? I have no clue what you try to say here :)

File: src/strings.cpp [hunk 4]
```diff
@@ -1625,38 +1618,34 @@ static char *FormatString(char *buff, const char *str_arg, StringParameters *arg
...
-	if ((x & FACIL_BUS_STOP)   && (buff + Utf8CharLen(SCC_BUS)   < last)) buff += Utf8Encode(buff, SCC_BUS);
-	if ((x & FACIL_DOCK)       && (buff + Utf8CharLen(SCC_SHIP)  < last)) buff += Utf8Encode(buff, SCC_SHIP);
-	if ((x & FACIL_AIRPORT)    && (buff + Utf8CharLen(SCC_PLANE) < last)) buff += Utf8Encode(buff, SCC_PLANE);
-	*buff = '\0';
-	return buff;
+	if (x & FACIL_TRAIN)      builder.Utf8Encode(SCC_TRAIN);
+	if (x & FACIL_TRUCK_STOP) builder.Utf8Encode(SCC_LORRY);
+	if (x & FACIL_BUS_STOP)   builder.Utf8Encode(SCC_BUS);
+	if (x & FACIL_DOCK)       builder.Utf8Encode(SCC_SHIP);
+	if (x & FACIL_AIRPORT)    builder.Utf8Encode(SCC_PLANE);
```
Comment (TrueBrain) [inline src/strings.cpp]: While at it, can we drop the silly alignment spaces here? :)

File: src/strings.cpp [hunk 5]
```diff
@@ -1625,38 +1618,34 @@ static char *FormatString(char *buff, const char *str_arg, StringParameters *arg
...
+static void StationGetSpecialString(StringBuilder &builder, int x)
 {
-	if ((x & FACIL_TRAIN)      && (buff + Utf8CharLen(SCC_TRAIN) < last)) buff += Utf8Encode(buff, SCC_TRAIN);
-	if ((x & FACIL_TRUCK_STOP) && (buff + Utf8CharLen(SCC_LORRY) < last)) buff += Utf8Encode(buff, SCC_LORRY);
-	if ((x & FACIL_BUS_STOP)   && (buff + Utf8CharLen(SCC_BUS)   < last)) buff += Utf8Encode(buff, SCC_BUS);
-	if ((x & FACIL_DOCK)       && (buff + Utf8CharLen(SCC_SHIP)  < last)) buff += Utf8Encode(buff, SCC_SHIP);
-	if ((x & FACIL_AIRPORT)    && (buff + Utf8CharLen(SCC_PLANE) < last)) buff += Utf8Encode(buff, SCC_PLANE);
-	*buff = '\0';
-	return buff;
+	if (x & FACIL_TRAIN)      builder.Utf8Encode(SCC_TRAIN);
```
Comment (TrueBrain) [inline src/strings.cpp]: While at it, I wonder if we should make those bit-checks explicit, as I think we normally do that? `((x & FACIL_XXX) != 0)` stuff.

File: src/strings_internal.h [hunk 6]
```diff
@@ -118,6 +118,17 @@ class StringBuilder {
...
 		return (ptrdiff_t)(this->last - this->current);
 	}
+
+	/**
+	 * Migration helper function to add a string using the old/to be deprecated
+	 * method of: new_destination = function(destination, last).
+	 * @param function The function to pass the current and last location to,
+	 *                 that will then return the new current location.
+	 */
+	void AddOldStyle(std::function<char*(char*, const char*)> function)
```
Comment (TrueBrain) [inline src/strings_internal.h]: I know you have the intention to remove this eventually, but the name is a bit odd. `AddCString`? I dunno .. but `Old` is not a word that is helping for context what this does :)

**Review by TrueBrain — COMMENTED** [review]
Nothing code-wise, all about the structure of things; sorry about that. But it is really hard to follow sometimes, when you are not deep into this stuff :)

File: src/strings.cpp (hunk 2 above)
Comment (rubidium42) [reply src/strings.cpp:850]: Added some comments in the code what is happening/why it is happening, but it that doesn't make it show as outdated here.

File: src/strings.cpp [hunk 7]
```diff
@@ -806,18 +828,26 @@ static std::vector<const char *> _game_script_raw_strings;
 
 /**
  * Parse most format codes within a string and write the result to a buffer.
- * @param buff    The buffer to write the final string to.
+ * @param builder The string builder to write the final string to.
  * @param str_arg The original string with format codes.
  * @param args    Pointer to extra arguments used by various string codes.
- * @param last    Pointer to just past the end of the buff array.
  * @param dry_run True when the argt array is not yet initialized.
```
Comment (TrueBrain) [inline src/strings.cpp]: lol; `argt`. Sorry, not your fault, but typo :P If there is nothing else, you can ignore this comment :)

File: src/strings.cpp [hunk 8]
```diff
@@ -806,18 +828,26 @@ static std::vector<const char *> _game_script_raw_strings;
...
  */
-static char *FormatString(char *buff, const char *str_arg, StringParameters *args, const char *last, uint case_index, bool game_script, bool dry_run)
+static void FormatString(StringBuilder &builder, const char *str_arg, StringParameters *args, uint case_index, bool game_script, bool dry_run)
 {
 	uint orig_offset = args->offset;
 
 	/* When there is no array with types there is no need to do a dry run. */
```
Comment (TrueBrain) [inline src/strings.cpp]: This comment .. seems wrong or outdated. I guess by now it is more what you wrote in the blob below.

`args` isn't initialized, and this wasn't a dry-run, so call ourselves again to do a dry-run, so `args` becomes initialized.

I would say, it is almost better to remove this comment, honestly.

**Review by TrueBrain — DISMISSED** [review]

**Review by TrueBrain — APPROVED** [review]
